llvm/lib/Support/Unix/Memory.inc

```diff
@@ -251,5 +320,50 @@ void Memory::InvalidateInstructionCache(const void *Addr, size_t Len) {
   ValgrindDiscardTranslations(Addr, Len);
 }
 
+static inline bool isPermissionError(int err) {
+  // PaX uses EPERM, SELinux uses EACCES
+  return err == EPERM || err == EACCES;
+}
+
+bool Memory::execProtectionChangeNeedsNewMapping() {
+#if defined(__linux__)
+  static int status = -1;
+
+  if (status != -1)
+    return status;
+
+  // Try to get the status from /proc/self/status, looking for PaX flags.
+  if (auto file = MemoryBuffer::getFileAsStream("/proc/self/status")) {
+    auto pax_flags =
+        (*file)->getBuffer().rsplit("PaX:").second.split('\n').first.trim();
+    if (!pax_flags.empty())
+      // 'M' indicates MPROTECT is enabled
+      return (status = pax_flags.find('M') != StringRef::npos);
+  }
+
+  // Create a temporary writable mapping and try to make it executable.  If
+  // this fails, test 'errno' to ensure it failed because we were not allowed
+  // to create such a mapping and not because of some transient error.
+  size_t size = Process::getPageSizeEstimate();
+  void *addr = ::mmap(NULL, size, PROT_READ | PROT_WRITE,
+                      MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+  if (addr == MAP_FAILED) {
+    // Must be low on memory or have too many mappings already, not much we can
+    // do here.
+    status = 0;
+  } else {
+    if (::mprotect(addr, size, PROT_READ | PROT_EXEC) < 0)
+      status = isPermissionError(errno);
```

Comment:
Oh, I wasn't aware that there's a mode where an "execmem" denial would kill the involved process. During my tests on Fedora I only observed failing syscalls returning `-EACCESS` (this very `mprotect()` call).

While I'd like to have a better detection of SELinux's "deny_execmem", there are basically two reasons against it:

1. If this sequence of syscalls (`addr = mmap(RW); mprotect(addr, RX)`) ends up killing the process instead of only making the second syscall fail, it probably shouldn't attempt to make use of a JIT anyways -- with a host configuration that strict.
2. Linking against `libselinux` by default just for this test is a deal breaker as we (a) shouldn't require all users of `libLLVM` to link against `libselinux` (well, technically we could but) and (b) overriding library internal state -- even if only briefly -- will mess with the application also trying to make use of `libselinux` and that'll lead to unintentional misbehavior for applications linking against `libLLVM`. A library shouldn't do that.

We can, however, add a mode (build time option) where we do the specific checks for distributions that do care and want to have such a narrow check. It should, however, try to make use of `libselinux` without changing any of its internal state.

Reply:
hmmph, I didn't think of the multithreaded issue of "one thread is the consumer setting a selinux callback, the other is the JIT doing it's thing". I'm not sure how to best solve this then, as I'd like to avoid having llvm print selinux log messages.

I guess at worst we could simply not check for `execmem` specifically and instead just bail out when we find selinux enabled? I'd argue that any sane selinux instance disabled `execmem` anyways :P

Reply:
> hmmph, I didn't think of the multithreaded issue of "one thread is the consumer setting a selinux callback, the other is the JIT doing it's thing". I'm not sure how to best solve this then, as I'd like to avoid having llvm print selinux log messages.

Yeah, me too. Having a library printing stuff at "arbitrary" times isn't nice ;)

> 
> I guess at worst we could simply not check for `execmem` specifically and instead just bail out when we find selinux enabled? I'd argue that any sane selinux instance disabled `execmem` anyways :P

From my experience, quite the opposite. You won't find a current desktop system that has SELinux's "deny_execmem" enabled because there's just so many JITs out there all trying to either directly `mmap(RWX)` or do the transition `RW-` -> `R-X`.

So, no, just because SELinux is enabled, doesn't mean its policy will deny generating code at runtime. That's why the test in `Memory::execProtectionChangeNeedsNewMapping()` has a generic fallback if a specific test (for PaX only, so far) fails.

Reply:
I (and I bet everyone else who works on selinux policies) would still much prefer being overly cautious rather than causing an AVC denial to detect `execmem` support.
It means we'd have to add a suppression for everything that uses the llvm jit, which means losing any information / audit logging on when a program unintentionally tries to use `execmem`.

Reply:
Got it. I just want to stress, the status quo would be to generate the very same AVC denial _and_ having the JIT code generation fail with `EACCESS` or even killing the process, as you've hinted before.

My changes do not make it worse, in fact, they make it better for configurations that do not kill a process on an AVC denial. But yeah, I see that preventing the AVC denial is a sensible goal as well.

I'll try to see what I can do about it without making libLLVM mess with libselinux's global state.

Reply:
Implementing a _freestanding_ check was actually relatively easy. ...well, kind of. I was able to implement a C version in basically no time but I struggled to make that comply with LLVM's set of C++ helpers around file i/o. In fact, I failed to find something that would suit my needs for opening a file as a stream for reading and writing. That's why I went for `std::fstream` instead. If that's a no-go, please, PLEASE, point me to some abstraction that's better suited. (Honestly, I even find the hops one has to go through just to make use of `MemoryBuffer::getFileAsStream()` ugly but beauty lies in the eye of the beholder, they say....). Just for the record, that would be the C pendant:

```C
#include <sys/stat.h>
#include <stdint.h>
#include <stdlib.h>
#include <stdio.h>

int is_selinux_enabled(void) __attribute__((weak));

#ifndef SEL_INO_MASK
#define SEL_INO_MASK 0x00ffffff
#endif
#ifndef SEL_CLASS_INO_OFFSET
#define SEL_CLASS_INO_OFFSET 0x04000000
#endif
#ifndef SEL_VEC_MAX
#define SEL_VEC_MAX 32
#endif
#ifndef SELINUX_MAGIC
#define SELINUX_MAGIC 0xf97cff8c
#endif

#define SEL_MNT "/sys/fs/selinux"

int selinux_can_execmem(void) {
    if (is_selinux_enabled && !is_selinux_enabled())
        return 1;

    // get current context
    char ctx[1024] = {};
    FILE *f = fopen("/proc/self/attr/current", "r");
    if (!f)
        return -1;
    if (!fgets(ctx, sizeof(ctx), f)) {
        fclose(f);
        return -1;
    }
    fclose(f);

    // The inode encodes class and permission bits
    struct stat sbuf;
    if (stat(SEL_MNT "/class/process/perms/execmem", &sbuf))
        return -1;

    if (!(sbuf.st_ino & SEL_CLASS_INO_OFFSET))
        return -1;

    int clazz = (sbuf.st_ino & SEL_INO_MASK) / (SEL_VEC_MAX + 1);
    int xperm = (sbuf.st_ino & SEL_INO_MASK) % (SEL_VEC_MAX + 1) - 1;
    if (xperm < 0 || xperm > 31)
        return -1;

    unsigned int execmem = 1U << xperm;

    f = fopen(SEL_MNT "/access", "r+");
    if (!f)
        return -1;

    if (fprintf(f, "%s %s %d %x", ctx, ctx, clazz, execmem) < 0) {
        fclose(f);
        return -1;
    }

    unsigned int lo, hi;
    if (fscanf(f, "%x %x", &lo, &hi) != 2) {
        fclose(f);
        return -1;
    }
    fclose(f);

    uint64_t allowed = (uint64_t)hi << 32 | lo;
    return !!(allowed & execmem);

}
int main(void) {
    int status = selinux_can_execmem();
    printf("I am %sallowed to execmem\n",
           status == 0 ? "not " :
           status == 1 ? "" :
           "maybe ");
    return 0;
}
```

Either way, the current version avoids the AVC denial by implementing a specific test which should address your concerns.